### PR TITLE
Add SameElementItem for struct search support.

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe or @arnej27959 PR
@bratseth FYI

### CODE REVIEW

File: container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.prelude.query;
+
+
+import com.yahoo.protect.Validator;
+
+import java.util.Iterator;
+
+/**
+ * This represents a query where all terms are required to match in the sma element id.
```
Comment (arnej27959) [inline container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java]: sma -> same

File: container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        return getItemCount();
+    }
+
+    @Override
+    protected void adding(Item item) {
+        Validator.ensureInstanceOf("Child item", item, TermItem.class);
+        TermItem asTerm = (TermItem) item;
+        Validator.ensureNotNull("Struct fieldname", asTerm.getIndexName());
+        Validator.ensureNotNull("Query term", asTerm.getIndexedString());
+        Validator.ensureNonEmpty("Struct fieldname", asTerm.getIndexName());
```
Comment (arnej27959) [inline container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java]: do you really need both NonNull and NonEmpty?

**Review by arnej27959 — APPROVED** [review]
note that we'll need to add YQL+ before this can actually be used, since tracing depends on it.

File: container-search/src/test/java/com/yahoo/prelude/query/test/SameElementItemTestCase.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import static org.junit.Assert.assertEquals;
+
+public class SameElementItemTestCase {
+    @Test
+    public void testAddItem() {
+        SameElementItem s = new SameElementItem("structa");
+        s.addItem(new WordItem("b", "f1"));
+        s.addItem(new WordItem("c", "f2"));
+        s.addItem(new WordItem("d", "f3"));
+        assertEquals("SAME_ELEMENT f1:b f2:c f3:d", s.toString());
```
Comment (havardpe) [inline container-search/src/test/java/com/yahoo/prelude/query/test/SameElementItemTestCase.java]: consider letting toString also mention the struct name

File: container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+        for (Iterator<Item> i = getItemIterator(); i.hasNext();) {
+            IndexedItem indexedItem = (IndexedItem) i.next();
+
+            buf.append(indexedItem.getIndexedString());
+            if (i.hasNext()) {
+                buf.append(' ');
+            }
+        }
+        return buf.toString();    }
```
Comment (havardpe) [inline container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java]: misplaced end brace

File: container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java (hunk 1 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java]: Fixed

File: container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java (hunk 2 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java]: You are right. index is guaranteed non-null in TermItem.
There also seem to be protection null terms in the various Items.

File: container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java (hunk 4 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/prelude/query/SameElementItem.java]: Fixed

File: container-search/src/test/java/com/yahoo/prelude/query/test/SameElementItemTestCase.java (hunk 3 above)
Comment (baldersheim) [reply container-search/src/test/java/com/yahoo/prelude/query/test/SameElementItemTestCase.java]: Fixed.

**Comment by baldersheim** [comment]
@havardpe and @arnej27959 All comments addressed.

**Pull request merged**